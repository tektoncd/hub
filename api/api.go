package apiserver

import (
	api "api/gen/api"
	"context"
	"log"
)

// api service example implementation.
// The example methods log the requests and return zero values.
type apisrvc struct {
	logger *log.Logger
}

// NewAPI returns the api service implementation.
func NewAPI(logger *log.Logger) api.Service {
	return &apisrvc{logger}
}

// Get all tasks and pipelines.
func (s *apisrvc) List(ctx context.Context) (err error) {
	s.logger.Print("api.list")
	return
}
