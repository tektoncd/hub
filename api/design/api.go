// Copyright Â© 2020 The Tekton Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package design

import (
	. "goa.design/goa/v3/dsl"
	cors "goa.design/plugins/v3/cors/dsl"

	// Enables the zaplogger plugin
	_ "goa.design/plugins/v3/zaplogger"
)

var _ = API("hub", func() {
	Title("Tekton Hub")
	Description("HTTP services for managing Tekton Hub")
	Version("0.1")
	Meta("swagger:example", "false")
	Server("hub", func() {
		Host("production", func() {
			URI("http://api.hub.tekton.dev")
		})

		Services(
			"admin",
			"auth",
			"category",
			"rating",
			"resource",
			"status",
			"swagger",
		)
	})

	// TODO: restrict CORS origin | https://github.com/tektoncd/hub/issues/26
	cors.Origin("*", func() {
		cors.Headers("Content-Type", "Authorization")
		cors.Methods("GET", "POST", "PUT")
	})
})
