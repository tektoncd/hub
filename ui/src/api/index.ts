import axios from 'axios';
import { API_URL } from '../config/constants';
import { ICategory } from '../store/category';
import { IResource, IVersion } from '../store/resource';

interface Token {
  token: string;
  refreshInterval: string;
  expiresAt: number;
}

interface TokenData {
  access: Token;
  refresh: Token;
}

export interface AuthResponse {
  data: TokenData;
}

export interface Rating {
  rating: number;
}

export interface Api {
  categories(): Promise<ICategory>;
  resources(): Promise<IResource>;
  resourceVersion(resourceId: number): Promise<IVersion>;
  versionUpdate(versionId: number): Promise<IVersion>;
  authentication(authCode: string): Promise<AuthResponse>;
  readme(rawURL: string): Promise<string>;
  yaml(rawURL: string): Promise<string>;
  getRating(resourceId: number, token: string): Promise<Rating>;
  setRating(resourceId: number, token: string, rating: number): Promise<void | null>;
}

export class Hub implements Api {
  async categories() {
    try {
      return axios.get(`${API_URL}/categories`).then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async resources() {
    try {
      return axios.get(`${API_URL}/resources`).then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async authentication(authCode: string) {
    try {
      return axios
        .post(`${API_URL}/auth/login?code=${authCode}`)
        .then((response) => response.data)
        .catch((err) => Promise.reject(err.response));
    } catch (err) {
      return err;
    }
  }

  async resourceVersion(resourceId: number) {
    try {
      return axios
        .get(`${API_URL}/resource/${resourceId}/versions`)
        .then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async versionUpdate(versionId: number) {
    try {
      return axios
        .get(`${API_URL}/resource/version/${versionId}`)
        .then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async readme(rawURL: string) {
    try {
      const URL = rawURL.substring(0, rawURL.lastIndexOf('/') + 1);
      return axios.get(`${URL}/README.md`).then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async yaml(rawURL: string) {
    try {
      const newLine = '\n';
      return axios.get(`${rawURL}`).then((response) => '```yaml' + newLine + response.data);
    } catch (err) {
      return err.response;
    }
  }

  async getRating(resourceId: number, token: string) {
    try {
      return axios
        .get(`${API_URL}/resource/${resourceId}/rating`, {
          headers: {
            Authorization: `Bearer ${token}`
          }
        })
        .then((response) => response.data)
        .catch((err) => Promise.reject(err.response));
    } catch (err) {
      return err;
    }
  }

  async setRating(resourceId: number, token: string, rating: number) {
    try {
      return axios
        .put(
          `${API_URL}/resource/${resourceId}/rating`,
          { rating: rating },
          {
            headers: {
              Authorization: `Bearer ${token}`
            }
          }
        )
        .then((response) => response.data)
        .catch((err) => Promise.reject(err.response));
    } catch (err) {
      return err;
    }
  }
}
