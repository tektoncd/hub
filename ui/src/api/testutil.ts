import { Api } from './';
import * as fs from 'fs';
import { ICategory } from '../store/category';
import { IResource, IVersion } from '../store/resource';

export class FakeHub implements Api {
  dataDir: string;

  constructor(dataDir: string) {
    this.dataDir = dataDir;
  }

  async categories() {
    const data = `${this.dataDir}/categories.json`;
    const ret = () => JSON.parse(fs.readFileSync(data).toString());
    return new Promise<ICategory>((resolve) => {
      setTimeout(() => resolve(ret()), 1000);
    });
  }

  async resources() {
    const data = `${this.dataDir}/resources.json`;

    const ret = () => JSON.parse(fs.readFileSync(data).toString());
    return new Promise<IResource>((resolve) => {
      setTimeout(() => resolve(ret()), 1000);
    });
  }

  async resourceVersion(resourceId: number) {
    const data = `${this.dataDir}/resource_versions_${resourceId}.json`;

    const ret = () => JSON.parse(fs.readFileSync(data).toString());
    return new Promise<IVersion>((resolve) => {
      setTimeout(() => resolve(ret()), 1000);
    });
  }

  async versionUpdate(versionId: number) {
    const data = `${this.dataDir}/version_details_${versionId}.json`;

    const ret = () => JSON.parse(fs.readFileSync(data).toString());
    return new Promise<IVersion>((resolve) => {
      setTimeout(() => resolve(ret()), 1000);
    });
  }
}
