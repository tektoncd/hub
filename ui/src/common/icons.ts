export enum Icons {
  Unknown = 'unknown',
  Cat = 'cat',
  Certificate = 'certificate',
  User = 'user',
  Build = 'build',
  Domain = 'domain'
}
