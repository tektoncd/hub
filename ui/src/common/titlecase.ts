/**
 * To convert the first letter of the input word to Upper Case
 */
export const titleCase = (filterName: string) => {
  return filterName.replace(/\w\S*/g, function (word) {
    return word.charAt(0).toUpperCase() + word.substr(1).toLowerCase();
  });
};
