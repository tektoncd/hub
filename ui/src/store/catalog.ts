import { Instance, types, flow, getEnv } from 'mobx-state-tree';
import { Icons } from '../common/icons';
import { titleCase } from '../common/titlecase';
import { Api } from '../api';

const icons: { [catalog: string]: Icons } = {
  community: Icons.Catalog
};

export const Catalog = types
  .model({
    id: types.identifierNumber,
    name: types.optional(types.string, ''),
    type: types.optional(types.string, ''),
    provider: types.optional(types.string, ''),
    selected: false
  })
  .actions((self) => ({
    toggle() {
      self.selected = !self.selected;
    }
  }))
  .views((self) => ({
    get icon(): Icons {
      return icons[self.type] || Icons.Catalog;
    }
  }));

export type ICatalog = Instance<typeof Catalog>;
export type ICatalogStore = Instance<typeof CatalogStore>;

export const CatalogStore = types
  .model({
    items: types.map(Catalog),
    isLoading: true,
    err: ''
  })

  .actions((self) => ({
    add(item: ICatalog) {
      self.items.put({ id: item.id, name: item.name, type: item.type, provider: item.provider });
    },
    clearSelected() {
      self.items.forEach((c) => {
        c.selected = false;
      });
    },
    toggleByName(name: string) {
      self.items.forEach((c) => {
        if (titleCase(c.name) === name) {
          c.selected = true;
        }
      });
    },
    setLoading(l: boolean) {
      self.isLoading = l;
    }
  }))

  .views((self) => ({
    get values() {
      return Array.from(self.items.values());
    },

    get api(): Api {
      return getEnv(self).api;
    },

    get selected() {
      const list = new Set();
      self.items.forEach((c: ICatalog) => {
        if (c.selected) {
          list.add(c.id);
        }
      });

      return list;
    },

    /* This view returns list of the selected catalos's name instead of id
    to avoid loop on it inorder to get catalogs name */
    get selectedByName() {
      return Array.from(self.items.values())
        .filter((c: ICatalog) => c.selected)
        .reduce((acc: string[], c: ICatalog) => [...acc, c.name], []);
    }
  }))

  .actions((self) => ({
    load: flow(function* () {
      try {
        self.setLoading(true);
        const { api } = self;

        const json = yield api.catalogs();

        const catalogs: ICatalog[] = json.data.map((c: ICatalog) => ({
          id: c.id,
          name: c.name,
          type: c.type,
          provider: c.provider
        }));

        catalogs.forEach((c: ICatalog) => self.add(c));
      } catch (err) {
        self.err = err.toString();
      }

      self.setLoading(false);
    })
  }))

  .actions((self) => ({
    afterCreate() {
      self.load();
    }
  }));
