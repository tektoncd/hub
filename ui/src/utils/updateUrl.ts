import { Params } from '../common/params';
import { SortByFields } from '../store/resource';
import { titleCase } from '../common/titlecase';

// This function returns all selected filters in a combination of params
export const UpdateURL = (
  search: string,
  sort: string,
  categories: string,
  kinds: string,
  catalogs: string,
  tags: Array<string>
) => {
  // To get URLSearchParams object instance
  const searchParams = new URLSearchParams(window.location.search);

  // To delete query params if search is empty
  if (!search && searchParams.has(Params.Query)) searchParams.delete(Params.Query);

  // Sets query params
  if (search) searchParams.set(Params.Query, search);

  //To delete sort params if doesn't selected any sort
  if (sort === SortByFields.Unknown && searchParams.has(Params.SortBy))
    searchParams.delete(Params.SortBy);

  // Set sort params
  if (sort !== SortByFields.Unknown) searchParams.set(Params.SortBy, sort);

  // To delete category params if doesn't selected any category
  if (!categories && searchParams.has(Params.Category)) searchParams.delete(Params.Category);

  // Sets category params
  if (categories) searchParams.set(Params.Category, categories);

  // To delete kind params if doesn't selected any kind
  if (!kinds && searchParams.has(Params.Kind)) searchParams.delete(Params.Kind);

  // Sets Kind params
  if (kinds) searchParams.set(Params.Kind, kinds);

  // To delete catalog params if doesn't selected any catalog
  if (!catalogs && searchParams.has(Params.Catalog)) searchParams.delete(Params.Catalog);

  // Sets catalos params
  if (catalogs) searchParams.set(Params.Catalog, titleCase(catalogs));

  // To delete tag param
  if (tags.length === 0) searchParams.delete(Params.Tag);

  // Sets tag params
  if (tags.length > 0) {
    searchParams.delete(Params.Query);
    searchParams.delete(Params.Tag);
    tags.forEach((t: string) => {
      searchParams.append(Params.Tag, t);
    });
  }

  return searchParams.toString();
};
