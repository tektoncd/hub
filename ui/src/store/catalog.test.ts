import { getSnapshot } from 'mobx-state-tree';
import { when } from 'mobx';
import { Catalog, CatalogStore } from './catalog';
import { assert } from './utils';
import { Icons } from '../common/icons';
import { FakeHub } from '../api/testutil';

const TESTDATA_DIR = `${__dirname}/testdata`;
const api = new FakeHub(TESTDATA_DIR);

describe('Store Object', () => {
  it('can create a catalog object', () => {
    const store = Catalog.create({
      id: 1,
      name: 'tekton',
      type: 'community'
    });

    expect(store.name).toBe('tekton');
  });

  it('creates a catalog store', (done) => {
    const store = CatalogStore.create({}, { api });
    expect(store.isLoading).toBe(true);
    when(
      () => !store.isLoading,
      () => {
        expect(store.items.size).toBe(2);
        expect(store.isLoading).toBe(false);

        expect(getSnapshot(store)).toMatchSnapshot();

        done();
      }
    );
  });

  it('should toggle a selected catalog', (done) => {
    const store = CatalogStore.create({}, { api });
    expect(store.isLoading).toBe(true);
    when(
      () => !store.isLoading,
      () => {
        expect(store.isLoading).toBe(false);
        const catalogs = store.items.get('1');
        assert(catalogs);
        catalogs.toggle();

        expect(store.selected.size).toBe(1);
        expect(catalogs.selected).toBe(true);

        done();
      }
    );
  });

  it('should clear all the selected catalog', (done) => {
    const store = CatalogStore.create({}, { api });
    when(
      () => !store.isLoading,
      () => {
        expect(store.isLoading).toBe(false);

        const catalogs = store.items.get('1');
        assert(catalogs);
        catalogs.toggle();

        store.clearSelected();

        expect(catalogs.selected).toBe(false);
        done();
      }
    );
  });

  it('can get an icon for catalog', (done) => {
    const store = CatalogStore.create({}, { api });

    when(
      () => !store.isLoading,
      () => {
        const catalog = store.items.get('1');
        assert(catalog);
        expect(catalog.icon).toBe(Icons.Catalog);

        done();
      }
    );
  });

  it('should toggle catalogs by name and can get selected catlogs by name', (done) => {
    const store = CatalogStore.create({}, { api });

    when(
      () => !store.isLoading,
      () => {
        store.toggleByName('Tekton');
        const catalogs = store.items.get('1');
        assert(catalogs);

        expect(catalogs.selected).toBe(true);
        expect(store.selectedByName).toEqual(['tekton']);

        done();
      }
    );
  });

  it('can get a provider for catalog', (done) => {
    const store = CatalogStore.create({}, { api });

    when(
      () => !store.isLoading,
      () => {
        const catalog = store.items.get('2');
        assert(catalog);
        expect(catalog.provider).toBe("gitlab");

        done();
      }
    );
  });
});
