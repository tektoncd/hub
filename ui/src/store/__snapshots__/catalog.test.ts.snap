// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store Object creates a catalog store 1`] = `
Object {
  "err": "",
  "isLoading": false,
  "items": Object {
    "1": Object {
      "id": 1,
      "name": "tekton",
      "provider": "github",
      "selected": false,
      "type": "community",
    },
    "2": Object {
      "id": 2,
      "name": "tekton-hub",
      "provider": "gitlab",
      "selected": false,
      "type": "official",
    },
  },
}
`;
